File: python/packages/handoffkit/handoffkit/browser/cache.py
"""Optional on-disk page cache."""

from __future__ import annotations

import hashlib
import json
import time
from pathlib import Path
from typing import Any

from handoffkit.browser.types import canonical_url


class BrowserCache:
    def __init__(self, root: str | Path = "", ttl_ms: int = 24 * 60 * 60 * 1000) -> None:
        self.root = Path(root).resolve() if root else Path()
        self.ttl_ms = int(ttl_ms)
        self.enabled = bool(root)

    def _path(self, url: str) -> Path:
        key = hashlib.sha256(canonical_url(url).encode("utf-8")).hexdigest()[:32]
        return self.root / f"{key}.json"

    def get(self, url: str) -> dict[str, Any] | None:
        if not self.enabled:
            return None
        path = self._path(url)
        try:
            raw = json.loads(path.read_text(encoding="utf-8"))
        except Exception:  # noqa: BLE001
            return None
        if not isinstance(raw, dict):
            return None
        if self.ttl_ms > 0 and time.time() * 1000 - float(raw.get("saved_at") or 0) > self.ttl_ms:
            return None
        return raw

    def set(self, url: str, payload: dict[str, Any] | None = None) -> bool:
        if not self.enabled:
            return False
        self.root.mkdir(parents=True, exist_ok=True)
        body = {"url": canonical_url(url), "saved_at": int(time.time() * 1000), **(payload or {})}
        self._path(url).write_text(json.dumps(body, ensure_ascii=False), encoding="utf-8")
        return True


def default_cache_root() -> Path:
    return Path.cwd() / ".cache" / "handoffkit-browser"

File: js/packages/browser/src/cache.js
import { createHash } from "node:crypto";
import { mkdir, readFile, writeFile } from "node:fs/promises";
import path from "node:path";
import { canonicalUrl } from "./util.js";

function hashKey(url) {
  return createHash("sha256").update(canonicalUrl(url)).digest("hex").slice(0, 32);
}

/** Optional on-disk cache for fetched page markdown / bodies. */
export class BrowserCache {
  constructor({ root = "", ttlMs = 24 * 60 * 60 * 1000 } = {}) {
    this.root = root ? path.resolve(root) : "";
    this.ttlMs = ttlMs;
    this.enabled = Boolean(this.root);
  }

  async get(url) {
    if (!this.enabled) return null;
    const file = path.join(this.root, `${hashKey(url)}.json`);
    try {
      const raw = JSON.parse(await readFile(file, "utf8"));
      if (!raw || typeof raw !== "object") return null;
      if (this.ttlMs > 0 && Date.now() - Number(raw.saved_at || 0) > this.ttlMs) return null;
      return raw;
    } catch {
      return null;
    }
  }

  async set(url, payload = {}) {
    if (!this.enabled) return false;
    await mkdir(this.root, { recursive: true });
    const file = path.join(this.root, `${hashKey(url)}.json`);
    const body = {
      url: canonicalUrl(url),
      saved_at: Date.now(),
      ...payload,
    };
    await writeFile(file, JSON.stringify(body), "utf8");
    return true;
  }
}

export function defaultCacheRoot() {
  return path.resolve(process.cwd(), ".cache", "handoffkit-browser");
}
